be/src/storage/binlog_file_writer.h

```diff
@@ -0,0 +1,197 @@
+};
+
+// Build the binlog file. A binlog file contains multiple pages, each page contains multiple log entries,
+// and each entry contains one or more change events(INSERT, UPDATE_BEFORE, UPDATE_AFTER, DELETE). Change
+// events from multiple ingestion are written to the file sequentially. The order is the same as that they
+// are committed. FE will assign a version for each ingestion when publishing, and each change event will
+// use this version to denote which ingestion they are generated in. A page only contains change events from
+// one version, and can not be shared among versions. The log entry is defined as a protobuf message, and
+// currently there are four types of log entries, and you can refer to binlog.proto for details.
+//    +--------------+------------------------------------------------------------------------------------------+
+//    | Type         | Description                                                                              |
+//    +==============+==========================================================================================+
+//    | INSERT_RANGE | Insert a batch of new rows into the table, and they are continuous in a segment file.    |
+//    |              | It will generate multiple INSERT change events                                           |
+//    +--------------+------------------------------------------------------------------------------------------+
+//    | UPDATE       | Update an existing row. It will generate a UPDATE_BEFORE and a UPDATE_AFTER.             |
+//    +--------------+------------------------------------------------------------------------------------------+
+//    | DELETE       | Delete an existing row. It will generate a DELETE change event                           |
+//    +--------------+------------------------------------------------------------------------------------------+
+//    | EMPTY        | There is no data change in an ingestion                                                        |
+//    +--------------+------------------------------------------------------------------------------------------+
+//
+// How to use
+//  std::shared_ptr<BinlogFileWriter> file_writer;
+//  file_writer->init();
+//  // prepare when starting to write data of a new version
+//  file_writer->begin();
+//  // add multiple items
+//  file_writer->add_insert_range():
+//  file_writer->add_update();
+//  file_writer->add_delete();
+//  // commit the binlog data in this version
+//  file_writer->commit();
+//  // abort this version if there is any error
+//  file_writer->abort();
+class BinlogFileWriter {
+public:
+    BinlogFileWriter(int64_t file_id, std::string path, int32_t page_size, CompressionTypePB compression_type);
+
+    // Initialize the file writer.
+    Status init();
+
+    // Begin to write binlog for a new version.
+    Status begin(int64_t version, const RowsetId& rowset_id, int64_t start_seq_id,
+                 int64_t change_event_timestamp_in_us);
```

Comment:
better not bind a version with single rowset_id, rowset_id info should be provided with logentry for each add method. that's more flexible, we have have multiple rowsets in single version in future.

Reply:
also there is a deltacolumn file design inprogress, I am not sure these 2 features are compatible

Reply:
1. remove rowset id from `begin`
2. for deltacolumn, need `<rowset_id, segment_index, row_id, version>` to find the data, the current file format contains these information